### PR TITLE
python3Packages.array-record: init at 0.4.0

### PR DESCRIPTION
**Pull request by GaetanLepage**
###### Description of changes

Add the [array_record](https://github.com/google/array_record) python package.
I used the `wheel` fetched from Pypi because the usual `python setup.py install` approach is not working (independently from Nix packaging).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Building the `.whl` from source involves messing with a bazel build.
See https://github.com/google/array_record/tree/main/oss.

File: pkgs/development/python-modules/array-record/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, absl-py
+, etils
+, importlib-resources
+, pythonOlder
+, typing-extensions
+, zipp
+}:
```
Comment (samuela) [inline pkgs/development/python-modules/array-record/default.nix:10]: nit: alphabetize

**Review by samuela — COMMENTED** [review]
The bazel source build would be preferable, but no objections to merging this as is

File: pkgs/development/python-modules/array-record/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    typing-extensions
+    zipp
+  ];
+
+  pythonImportsCheck = [ "array_record" ];
+
+  meta = with lib; {
+    description = "ArrayRecord is a new file format derived from Riegeli, achieving a new frontier of IO efficiency";
+    homepage = "https://github.com/google/array_record";
+    license = licenses.asl20;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/array-record/default.nix:40]: Do we need to add ``sourceProvenance = with sourceTypes; [ binaryNativeCode ];``?

File: pkgs/development/python-modules/array-record/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  version = "0.4.0";
+  format = "wheel";
+
+  disabled = pythonOlder "3.7";
+
+  src = fetchPypi {
+    inherit version format;
+    pname = "array_record";
+    dist = "py310";
+    python = "py310";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/array-record/default.nix:23]: This does not work for python 3.11 or pypy. Not sure if python has attrs for that.

Comment (GaetanLepage) [reply pkgs/development/python-modules/array-record/default.nix:23]: The binaries for python 3.11 are not distributed.
I have set `disabled = pythonVersion != "3.10".

File: pkgs/development/python-modules/array-record/default.nix (hunk 2 above)
Comment (GaetanLepage) [reply pkgs/development/python-modules/array-record/default.nix:40]: I have no idea :/

Comment (SuperSandro2000) [reply pkgs/development/python-modules/array-record/default.nix:40]: I am going to fix this tree wide in the builder function.

**Comment by SuperSandro2000** [comment]
@ofborg build python310Packages.array-record

**Pull request merged**